openrtb_ext/deal_tier.go

```diff
@@ -0,0 +1,68 @@
+package openrtb_ext
+
+import (
+	"encoding/json"
+	"strings"
+
+	"github.com/mxmCherry/openrtb"
+)
+
+// DealTier defines the configuration of a deal tier.
+type DealTier struct {
+	// Prefix specifies the beginning of the hb_pb_cat_dur targeting key value. Must be non-empty.
+	Prefix string `json:"prefix"`
+
+	// MinDealTier specifies the minimum deal priority value (inclusive) that must be met for the targeting
+	// key value to be modified. Must be greater than 0.
+	MinDealTier int `json:"minDealTier"`
+}
+
+// DealTierBidderMap defines a correlation between bidders and deal tiers.
+type DealTierBidderMap map[BidderName]DealTier
+
+// ReadDealTiersFromImp returns a map of bidder deal tiers read from the impression of an original request (not split / cleaned).
+func ReadDealTiersFromImp(imp openrtb.Imp) (DealTierBidderMap, error) {
+	dealTiers := make(DealTierBidderMap)
+
+	if len(imp.Ext) == 0 {
+		return dealTiers, nil
+	}
+
+	// imp.ext.{bidder}
+	var impExt map[string]struct {
+		DealTier *DealTier `json:"dealTier"`
+	}
+	if err := json.Unmarshal(imp.Ext, &impExt); err != nil {
+		return nil, err
+	}
+	for bidder, param := range impExt {
+		if param.DealTier != nil {
+			dealTiers[BidderName(bidder)] = normalizeDealTier(param.DealTier)
+		}
+	}
+
+	// imp.ext.prebid.{bidder}
+	var impPrebidExt struct {
+		Prebid struct {
+			Bidders map[string]struct {
+				DealTier *DealTier `json:"dealTier"`
+			} `json:"bidder"`
+		} `json:"prebid"`
+	}
+	if err := json.Unmarshal(imp.Ext, &impPrebidExt); err != nil {
+		return nil, err
+	}
+	for bidder, param := range impPrebidExt.Prebid.Bidders {
+		if param.DealTier != nil {
+			dealTiers[BidderName(bidder)] = normalizeDealTier(param.DealTier)
+		}
+	}
+
+	return dealTiers, nil
+}
+
+func normalizeDealTier(dealTier *DealTier) DealTier {
```

Comment:
Was there a spec or an identified use case where normalization would be needed?

Reply:
I don't think there is a spec that this is needed. It was originally added as some basic additional protection against bad configurations. As we are adding updates to the docs, we could either add a note about what makes a proper prefix, or perhaps this function could be removed if we don't want to enforce this.

Reply:
I prefer to remove the logic and update the docs. In general, PBS does not historically protect the publisher from bad config by silently trying to fix it. We like to fail loud and fail fast.

Sounds good @hhhjort @camrice ?

Reply:
If it is a bad configuration that would never work, the whole request should be rejected, so the publisher will fix the config immediately rather than limping along in a suboptimal or unintended configuration. If the error is such that a config change on the PBS host determines if the config is valid, then we just warn so that a PBS config update does not break existing configs that were working.

Reply:
I agree with removing this for this PR and replacing with a warning if possible. That can be a future enhancement as well.

Reply:
Sounds good. I'll remove it. We don't have a good way of sending back warnings at the moment. I'll create an internal ticket for it.